enterprise/server/cmd/action_runner/main.go

```diff
@@ -0,0 +1,662 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net/url"
+	"os"
+	"os/exec"
+	"os/user"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/enterprise/server/workflowconf"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/google/shlex"
+	"github.com/google/uuid"
+	"google.golang.org/grpc"
+	"gopkg.in/yaml.v2"
+
+	bespb "github.com/buildbuddy-io/buildbuddy/proto/build_event_stream"
+	bepb "github.com/buildbuddy-io/buildbuddy/proto/build_events"
+	pepb "github.com/buildbuddy-io/buildbuddy/proto/publish_build_event"
+)
+
+const (
+	// Name of the dir into which the repo is cloned.
+	repoDirName = "repo-root"
+	// Path where we expect to find actions config, relative to the repo root.
+	actionsConfigPath = "buildbuddy.yaml"
+
+	// Env vars
+	// NOTE: These env vars are not populated for non-private repos.
+	// TODO: Allow populating BUILDBUDDY_API_KEY for private repos,
+	// so that workflow invocations can be private.
+
+	repoUserEnvVarName  = "REPO_USER"
+	repoTokenEnvVarName = "REPO_TOKEN"
+
+	// Exit code placeholder used when a command doesn't return an exit code on its own.
+	noExitCode = -1
+	// "Unique"-ish exit code that indicates the runner should be retried
+	// due to an error that may be temporary.
+	retryableExitCode = 21
+
+	// progressFlushInterval specifies how often we should flush
+	// each Bazel command's output while it is running.
+	progressFlushInterval = 250 * time.Millisecond
+
+	// Webhook event names
+
+	pushEventName        = "push"
+	pullRequestEventName = "pull_request"
+
+	// ANSI codes for nicer output
+
+	textGreen  = "\033[32m"
+	textYellow = "\033[33m"
+	textBlue   = "\033[34m"
+	textPurple = "\033[35m"
+	textCyan   = "\033[36m"
+	textGray   = "\033[90m"
+	textReset  = "\033[0m"
+)
+
+var (
+	besBackend    = flag.String("bes_backend", "", "gRPC endpoint for BuildBuddy's BES backend.")
+	besResultsURL = flag.String("bes_results_url", "", "URL prefix for BuildBuddy invocation URLs.")
+	repoURL       = flag.String("repo_url", "", "URL of the Git repo to check out.")
+	commitSHA     = flag.String("commit_sha", "", "SHA of the commit to be checked out.")
+	triggerEvent  = flag.String("trigger_event", "", "Event type that triggered the action runner.")
+	triggerBranch = flag.String("trigger_branch", "", "Branch to check action triggers against.")
+
+	emptyEnv = map[string]string{}
```

Comment:
if you're just reading from this in runCommand, you can just pass nil instead -- ranging over a nil map is A-OK.